File: smash-ranks-android/app/src/main/java/com/garpr/android/activities/BaseActivity.java
package com.garpr.android.activities;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.v4.app.NavUtils;
import android.support.v4.app.TaskStackBuilder;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;

import com.garpr.android.App;
import com.garpr.android.R;
import com.garpr.android.misc.Heartbeat;
import com.garpr.android.misc.RegionManager.RegionHandle;
import com.garpr.android.misc.Timber;
import com.garpr.android.preferences.GeneralPreferenceStore;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

public abstract class BaseActivity extends AppCompatActivity implements Heartbeat, RegionHandle {

    private static final String TAG = "BaseActivity";
    private static final String CNAME = BaseActivity.class.getCanonicalName();
    protected static final String EXTRA_REGION = CNAME + ".Region";

    private Unbinder mUnbinder;

    @Inject
    protected GeneralPreferenceStore mGeneralPreferenceStore;

    @Inject
    protected Timber mTimber;

    @Nullable
    @BindView(R.id.toolbar)
    protected Toolbar mToolbar;


    protected abstract String getActivityName();

    @Nullable
    @Override
    public String getCurrentRegion() {
        final Intent intent = getIntent();

        if (intent != null && intent.hasExtra(EXTRA_REGION)) {
            return intent.getStringExtra(EXTRA_REGION);
        } else {
            return null;
        }
    }

    @Nullable
    protected CharSequence getSubtitle() {
        final ActionBar actionBar = getSupportActionBar();
        return actionBar == null ? null : actionBar.getSubtitle();
    }

    @Override
    public boolean isAlive() {
        return !isFinishing() && !isDestroyed();
    }

    protected void navigateUp() {
        final Intent upIntent = NavUtils.getParentActivityIntent(this);

        if (upIntent == null) {
            supportFinishAfterTransition();
        } else if (NavUtils.shouldUpRecreateTask(this, upIntent)) {
            TaskStackBuilder.create(this)
                    .addNextIntentWithParentStack(upIntent)
                    .startActivities();
        } else {
            supportNavigateUpTo(upIntent);
        }
    }

    @Override
    protected void onCreate(@Nullable final Bundle savedInstanceState) {
        App.get().getAppComponent().inject(this);
        // noinspection ConstantConditions
        getDelegate().setLocalNightMode(mGeneralPreferenceStore.getNightMode().get().getThemeValue());

        super.onCreate(savedInstanceState);
        mTimber.d(TAG, getActivityName() + " created");
    }

    @Override
    public boolean onOptionsItemSelected(final MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                navigateUp();
                return true;
        }

        return super.onOptionsItemSelected(item);
    }

    protected void onViewsBound() {
        if (mUnbinder != null) {
            mUnbinder.unbind();
        }

        mUnbinder = ButterKnife.bind(this);

        if (mToolbar != null) {
            setSupportActionBar(mToolbar);
        }

        final ActionBar actionBar = getSupportActionBar();

        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(showUpNavigation());
        }
    }

    @Override
    public void setContentView(@LayoutRes final int layoutResID) {
        super.setContentView(layoutResID);
        onViewsBound();
    }

    public void setSubtitle(@Nullable final CharSequence subtitle) {
        final ActionBar actionBar = getSupportActionBar();

        if (actionBar != null) {
            actionBar.setSubtitle(subtitle);
        }
    }

    public void setSubtitle(@StringRes final int resId) {
        setSubtitle(getText(resId));
    }

    protected boolean showUpNavigation() {
        return false;
    }

    @Override
    public String toString() {
        return getActivityName();
    }

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/activities/BaseActivity.kt
package com.garpr.android.activities

import android.os.Bundle
import android.view.MenuItem
import androidx.annotation.LayoutRes
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.TaskStackBuilder
import com.garpr.android.data.models.Region
import com.garpr.android.extensions.appComponent
import com.garpr.android.managers.RegionManager.RegionHandle
import com.garpr.android.misc.Heartbeat
import com.garpr.android.misc.Timber
import com.garpr.android.preferences.GeneralPreferenceStore
import javax.inject.Inject

abstract class BaseActivity : AppCompatActivity(), Heartbeat, RegionHandle {

    @Inject
    protected lateinit var generalPreferenceStore: GeneralPreferenceStore

    @Inject
    protected lateinit var timber: Timber


    companion object {
        private const val TAG = "BaseActivity"
        private val CNAME = BaseActivity::class.java.canonicalName
        internal val EXTRA_REGION = "$CNAME.Region"
    }

    protected abstract val activityName: String

    override val currentRegion: Region?
        get() = intent?.getParcelableExtra(EXTRA_REGION)

    override val isAlive: Boolean
        get() = !isFinishing && !isDestroyed

    open fun navigateUp() {
        val intent = supportParentActivityIntent

        if (intent == null) {
            supportFinishAfterTransition()
        } else if (supportShouldUpRecreateTask(intent)) {
            TaskStackBuilder.create(this)
                    .addNextIntentWithParentStack(intent)
                    .startActivities()
        } else {
            supportNavigateUpTo(intent)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        appComponent.inject(this)

        generalPreferenceStore.nightMode.get()?.let {
            delegate.setLocalNightMode(it.themeValue)
        } ?: throw RuntimeException("nightMode is null")

        super.onCreate(savedInstanceState)
        timber.d(TAG, "$activityName created")
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                navigateUp()
                true
            }

            else -> {
                super.onOptionsItemSelected(item)
            }
        }
    }

    protected open fun onViewsBound() {
        // intentionally empty, children can override
    }

    override fun setContentView(@LayoutRes layoutResID: Int) {
        super.setContentView(layoutResID)
        onViewsBound()
    }

    override fun toString() = activityName

}
